server/src/platform/drbg/mod.rs

```diff
@@ -0,0 +1,128 @@
+/// by SeedableRng for RNGs with such large seeds, the newtype pattern
+/// can be used:
+///
+/// ```rust
+/// const N: usize = 64;
+/// pub struct MyRngSeed(pub [u8; N]);
+///
+/// impl Default for MyRngSeed {
+///     fn default() -> MyRngSeed {
+///         MyRngSeed([0; N])
+///     }
+/// }
+///
+/// impl AsMut<[u8]> for MyRngSeed {
+///     fn as_mut(&mut self) -> &mut [u8] {
+///         &mut self.0
+///     }
+/// }
+/// ```
+///
+/// [`fill_bytes`]: Drbg::fill_bytes
+/// [`next_u32`]: Drbg::next_u32
+/// [`next_u64`]: Drbg::next_u64
+pub trait Drbg: Sized {
+    type Error: Debug;
+    /// Entropy type, which is restricted to types mutably-dereferenceable as u8 arrays
+    /// (we recommend either `[u8; N]` for some `N` or `GenericArray<u8, N>`).
+    type Entropy: Sized + Default + AsMut<[u8]>;
+    /// Nonce type, which is restricted to types mutably-dereferenceable as u8 arrays
+    /// (we recommend either `[u8; N]` for some `N` or `GenericArray<u8, N>`).
+    ///
+    /// According to 8.6.7: A nonce may be composed of one (or more) of the following
+    /// components (other components may also be appropriate):
+    ///
+    /// 1. A random value that is generated anew for each nonce, using an approved random bit
+    ///    generator.
+    /// 2. A timestamp of sufficient resolution (detail) so that it is different each time it
+    ///    is used.
+    /// 3. A monotonically increasing sequence number.
+    /// 4. A combination of a timestamp and a monotonically increasing sequence number, such
+    ///    that the sequence number is reset when and only when the timestamp changes. For
+    ///    example, a timestamp may show the date but not the time of day, so a sequence
+    ///    number is appended that will not repeat during a particular day.
+    ///
+    /// For case 1 above, the random value could be acquired from the same source and at the same time
+    /// as the entropy input. For case 2 above, the timestamp must be trusted. A trusted timestamp is
+    /// generated and signed by an entity that is trusted to provide accurate time information.
+    type Nonce: Sized + Default + AsMut<[u8]>;
+    /// Create a new DRBG using the given seed.
+    ///
+    /// All PRNG implementations should be reproducible: given a fixed seed,
+    /// the same sequence of output should be produced on all runs, library
+    /// versions and architectures (e.g. check endianness).
+    fn instantiate(
+        entropy_input: &Self::Entropy,
+        nonce: &Self::Nonce,
+        personalization_string: &[u8],
+    ) -> impl Drbg;
+    /// Add additional entropy to th current RNG
+    ///
+    /// All PRNG implementations should be reproducible: given a fixed seed,
+    /// the same sequence of output should be produced on all runs, library
+    /// versions and architectures (e.g. check endianness).
+    fn reseed(&mut self, entropy_input: &Self::Entropy, additional_input: &[u8]);
```

Comment:
I worry that it's a bit over-stating of our trait bounds to make `Entropy` an associated type. We anyways depend on our ability to get a slice out of it, why not just tell the caller "give me a slice of entropy"?

Personally I'd do that, and remove the `Entropy` associated type.

Reply:
my intention was to have drbg that cannot fail, but once we assume failure is ok, I think your suggestion makes lots of sense.

Reply:
Another reason is that certain drbgs have certain Nonce/ Entropy size requirements and those requirements are offloaded to the associated type, but can revisit this part later.